src/ngx_http_lua_shdict.c

```diff
@@ -503,7 +546,7 @@ ngx_http_lua_shdict_get_helper(lua_State *L, int get_stale)
                    "fetching key \"%V\" in shared dict \"%V\"", &key, &name);
 #endif /* NGX_DEBUG */
 
-    ngx_shmtx_lock(&ctx->shpool->mutex);
+    ctx->wlock(ctx);
```

Comment:
This defeats the benefit of the rwlock when looking up the key over huge rbtrees. I think we need finer granularity of locking. For the rbtree lookup, it should be reader lock while the LRU queue updating, it should be the writer lock.

Reply:
But for changing the rbtree (like updating nodes, deleting nodes, inserting nodes), then we need a write lock.